### PR TITLE
fix(collate): make optional dependency imports localized

### LINKED ISSUE #350: Bug: Importing pycytominer without cytominer_database installed fails
### Example code with output

```
TEST START: /home/conda/feedstock_root/build_artifacts/noarch/pycytominer-1.0.0-pyhd8ed1ab_0.conda
Renaming work directory '/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/work' to '/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/work_moved_pycytominer-1.0.0-pyhd8ed1ab_0_noarch'
INFO:conda_build.utils:Renaming work directory '/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/work' to '/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/work_moved_pycytominer-1.0.0-pyhd8ed1ab_0_noarch'
shutil.move(work)=/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/work, dest=/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/work_moved_pycytominer-1.0.0-pyhd8ed1ab_0_noarch)
INFO:conda_build.utils:shutil.move(work)=/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/work, dest=/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/work_moved_pycytominer-1.0.0-pyhd8ed1ab_0_noarch)
file:///home/conda/feedstock_root/build_artifact.. Checked  0.0s
file:///home/conda/feedstock_root/build_artifact.. Checked  0.0s
warning  libmamba Cache file "/home/conda/feedstock_root/build_artifacts/pkg_cache/cache/497deca9.json" was modified by another program
warning  libmamba Cache file "/home/conda/feedstock_root/build_artifacts/pkg_cache/cache/09cdf8bf.json" was modified by another program
file:///home/conda/feedstock_root/build_artifact..   1.3kB @  20.2MB/s  0.0s
file:///home/conda/feedstock_root/build_artifact.. 778.0 B @  16.2MB/s  0.0s
conda-forge/noarch                                  12.4MB @  22.1MB/s  0.8s
conda-forge/linux-64                                30.2MB @  17.3MB/s  2.1s
Reloading output folder: /home/conda/feedstock_root/build_artifacts
file:///home/conda/feedstock_root/build_artifact.. Checked  0.0s
file:///home/conda/feedstock_root/build_artifact.. Checked  0.0s
file:///home/conda/feedstock_root/build_artifact..   1.3kB @  18.5MB/s  0.0s
file:///home/conda/feedstock_root/build_artifact.. 778.0 B @  16.9MB/s  0.0s

## Package Plan ##

  environment location: /home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac


The following NEW packages will be INSTALLED:

    _libgcc_mutex:      0.1-conda_forge             conda-forge
    _openmp_mutex:      4.5-2_gnu                   conda-forge
    aws-c-auth:         0.7.5-h1a24852_0            conda-forge
    aws-c-cal:          0.6.7-h6e18cf3_0            conda-forge
    aws-c-common:       0.9.4-hd590300_0            conda-forge
    aws-c-compression:  0.2.17-h037bafe_4           conda-forge
    aws-c-event-stream: 0.3.2-he4fbe49_4            conda-forge
    aws-c-http:         0.7.13-hbbfb9a7_7           conda-forge
    aws-c-io:           0.13.35-hd1885a1_4          conda-forge
    aws-c-mqtt:         0.9.8-h31a96f8_0            conda-forge
    aws-c-s3:           0.3.20-he249171_1           conda-forge
    aws-c-sdkutils:     0.1.12-h037bafe_3           conda-forge
    aws-checksums:      0.1.17-h037bafe_3           conda-forge
    aws-crt-cpp:        0.24.4-h28e6ea9_2           conda-forge
    aws-sdk-cpp:        1.11.182-hb97d603_2         conda-forge
    bzip2:              1.0.8-h7f98852_4            conda-forge
    c-ares:             1.20.1-hd590300_1           conda-forge
    ca-certificates:    2023.7.22-hbcca054_0        conda-forge
    gflags:             2.2.2-he1b5a44_1004         conda-forge
    glog:               0.6.0-h6f12383_0            conda-forge
    greenlet:           3.0.1-py312h30efb56_0       conda-forge
    joblib:             1.3.2-pyhd8ed1ab_0          conda-forge
    keyutils:           1.6.1-h166bdaf_0            conda-forge
    krb5:               1.21.2-h659d440_0           conda-forge
    ld_impl_linux-64:   2.40-h41732ed_0             conda-forge
    libabseil:          20230802.1-cxx17_h59595ed_0 conda-forge
    libarrow:           13.0.0-hecbb4c5_13_cpu      conda-forge
    libblas:            3.9.0-19_linux64_openblas   conda-forge
    libbrotlicommon:    1.1.0-hd590300_1            conda-forge
    libbrotlidec:       1.1.0-hd590300_1            conda-forge
    libbrotlienc:       1.1.0-hd590300_1            conda-forge
    libcblas:           3.9.0-19_linux64_openblas   conda-forge
    libcrc32c:          1.1.2-h9c3ff4c_0            conda-forge
    libcurl:            8.4.0-hca28451_0            conda-forge
    libedit:            3.1.20191231-he28a2e2_2     conda-forge
    libev:              4.33-h516909a_1             conda-forge
    libevent:           2.1.12-hf998b51_1           conda-forge
    libexpat:           2.5.0-hcb278e6_1            conda-forge
    libffi:             3.4.2-h7f98852_5            conda-forge
    libgcc-ng:          13.2.0-h807b86a_2           conda-forge
    libgfortran-ng:     13.2.0-h69a702a_2           conda-forge
    libgfortran5:       13.2.0-ha4646dd_2           conda-forge
    libgomp:            13.2.0-h807b86a_2           conda-forge
    libgoogle-cloud:    2.12.0-h19a6dae_3           conda-forge
    libgrpc:            1.58.2-he06187c_0           conda-forge
    liblapack:          3.9.0-19_linux64_openblas   conda-forge
    libnghttp2:         1.55.1-h47da74e_0           conda-forge
    libnsl:             2.0.1-hd590300_0            conda-forge
    libnuma:            2.0.16-h0b41bf4_1           conda-forge
    libopenblas:        0.3.24-pthreads_h413a1c8_0  conda-forge
    libprotobuf:        4.24.3-hf27288f_1           conda-forge
    libre2-11:          2023.06.02-h7a70373_0       conda-forge
    libsqlite:          3.44.0-h2797004_0           conda-forge
    libssh2:            1.11.0-h0841786_0           conda-forge
    libstdcxx-ng:       13.2.0-h7e041cc_2           conda-forge
    libthrift:          0.19.0-hb90f79a_1           conda-forge
    libutf8proc:        2.8.0-h166bdaf_0            conda-forge
    libuuid:            2.38.1-h0b41bf4_0           conda-forge
    libzlib:            1.2.13-hd590300_5           conda-forge
    lz4-c:              1.9.4-hcb278e6_0            conda-forge
    ncurses:            6.4-h59595ed_2              conda-forge
    numpy:              1.26.0-py312heda63a1_0      conda-forge
    openssl:            3.1.4-hd590300_0            conda-forge
    orc:                1.9.0-h208142c_3            conda-forge
    pandas:             2.1.2-py312hfb8ada1_0       conda-forge
    pyarrow:            13.0.0-py312h176e3d2_13_cpu conda-forge
    pycytominer:        1.0.0-pyhd8ed1ab_0          local      
    python:             3.12.0-hab00c5b_0_cpython   conda-forge
    python-dateutil:    2.8.2-pyhd8ed1ab_0          conda-forge
    python-tzdata:      2023.3-pyhd8ed1ab_0         conda-forge
    python_abi:         3.12-4_cp312                conda-forge
    pytz:               2023.3.post1-pyhd8ed1ab_0   conda-forge
    rdma-core:          28.9-h59595ed_1             conda-forge
    re2:                2023.06.02-h2873b5e_0       conda-forge
    readline:           8.2-h8228510_1              conda-forge
    s2n:                1.3.55-h06160fa_0           conda-forge
    scikit-learn:       1.3.2-py312h394d371_1       conda-forge
    scipy:              1.11.3-py312heda63a1_1      conda-forge
    setuptools:         68.2.2-pyhd8ed1ab_0         conda-forge
    six:                1.16.0-pyh6c4a22f_0         conda-forge
    snappy:             1.1.10-h9fff704_0           conda-forge
    sqlalchemy:         1.4.49-py312h98912ed_1      conda-forge
    threadpoolctl:      3.2.0-pyha21a80b_0          conda-forge
    tk:                 8.6.13-h2797004_0           conda-forge
    tzdata:             2023c-h71feb2d_0            conda-forge
    ucx:                1.15.0-h64cca9d_0           conda-forge
    xz:                 5.2.6-h166bdaf_0            conda-forge
    zstd:               1.5.5-hfc55251_0            conda-forge

Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
export PREFIX=/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac
export SRC_DIR=/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/test_tmp
import: 'pycytominer'
Traceback (most recent call last):
  File "/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/test_tmp/run_test.py", line 2, in <module>
    import pycytominer
  File "/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/python3.12/site-packages/pycytominer/__init__.py", line 1, in <module>
    from .aggregate import aggregate
  File "/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/python3.12/site-packages/pycytominer/aggregate.py", line 7, in <module>
    from pycytominer.cyto_utils import (
  File "/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/python3.12/site-packages/pycytominer/cyto_utils/__init__.py", line 43, in <module>
    from .collate import collate
  File "/home/conda/feedstock_root/build_artifacts/pycytominer_1698862705829/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_plac/lib/python3.12/site-packages/pycytominer/cyto_utils/collate.py", line 6, in <module>
    import cytominer_database.ingest
ModuleNotFoundError: No module named 'cytominer_database'
Tests failed for pycytominer-1.0.0-pyhd8ed1ab_0.conda - moving package to /home/conda/feedstock_root/build_artifacts/broken
WARNING:conda_build.build:Tests failed for pycytominer-1.0.0-pyhd8ed1ab_0.conda - moving package to /home/conda/feedstock_root/build_artifacts/broken
TESTS FAILED: pycytominer-1.0.0-pyhd8ed1ab_0.conda
Traceback (most recent call last):
  File "/workspaces/pycytominer-feedstock/./build-locally.py", line 105, in <module>
    main()
  File "/workspaces/pycytominer-feedstock/./build-locally.py", line 93, in main
    run_docker_build(ns)
  File "/workspaces/pycytominer-feedstock/./build-locally.py", line 29, in run_docker_build
    subprocess.check_call([script])
  File "/opt/conda/lib/python3.11/subprocess.py", line 413, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['.scripts/run_docker_build.sh']' returned non-zero exit status 1.
```

### Issue description

This bug was discovered while packaging pycytominer for conda. The test `import pycytominer` fails for the default package, because a series of obligate imports requires that collate (and it's dependencies be present). 

This can likely be solved by getting rid of the imports in `__init__.py`. 

### Expected behavior

`import pycytominer` should work even when cytominer_database isn't installed. 

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by kenibrewer**
# Description

This change is required to allow 'import pycytominer' without triggering an error from missing dependencies. 

Closes #350

## What is the nature of your change?

- [x] Bug fix (fixes an issue).

# Checklist

Please ensure that all boxes are checked before indicating that a pull request is ready for review.

- [x] I have read the [CONTRIBUTING.md](CONTRIBUTING.md) guidelines.
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [ ] New and existing unit tests pass locally with my changes.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have deleted all non-relevant text in this pull request template.

### CODE REVIEW

**Comment by kenibrewer** [comment]
@d33bs I'm not sure if this is the most pythonic way to resolve this issue, but I wasn't able to find any consensus in my searching for a best practice approach to resolving it. Definitely open to suggestions.

File: pycytominer/cyto_utils/__init__.py
Comment (d33bs) [inline pycytominer/cyto_utils/__init__.py:49]: Instead of redefining `collate` here, consider moving the `cytominer_database` import within `collate.py` to a localized function import where it's used. This could grant a similar functionality without modifying the `__init__.py` or implementing redefinition of `collate`. [We already do this with the SingleCells class](https://github.com/cytomining/pycytominer/blob/main/pycytominer/cyto_utils/collate.py#L73) so there's some precedence to the style of the approach here.

When it comes to Pythonic approach, this is totally single opinion and not necessarily canon 🙂 . Generally I feel the above might be "simple over complex", avoid broad-reaching changes within the `__init__.py` (keeping impact minimized), and could help with avoiding future code decay by aligning logic closer to implementation.

File: pycytominer/cyto_utils/__init__.py
Comment (d33bs) [inline pycytominer/cyto_utils/__init__.py:44]: Partially about this line and more: depending on what you think about the other comment, to be extra thorough it might be a good idea to keep this comment (or something like it) around the function-local import so anyone looking at it understands the necessity of the import location and avoids what might look like "import organization cleanup work".

This brings up a different point about a fork in the road with test cases we have at the moment from source vs conda. To avoid this issue in the future ahead of a conda release should we try the very same test found in the conda recipe (i.e. `pip install pycytominer; import pycytominer`)? This might be considered ["shifting left on testing"](https://en.wikipedia.org/wiki/Shift-left_testing) here, detecting things earlier in the process with theoretical related benefits.

**Review by d33bs — COMMENTED** [review]
Nice job @kenibrewer ! I left a few comments which I wanted to get your insight on before an approval. Please don't hesitate to let me know if you have any questions (or if what I mentioned seems like a bad idea!).

File: pycytominer/cyto_utils/__init__.py
Comment (kenibrewer) [reply pycytominer/cyto_utils/__init__.py:44]: I definitely like the shift left on testing approach. I think it would be good to define which tests require which optional dependencies and configure tests to run in minimal environments. The one concern I would have would be about the testing matrix getting unwieldly large, but it might also mean faster more parallelized testing...

File: pycytominer/cyto_utils/__init__.py
Comment (kenibrewer) [reply pycytominer/cyto_utils/__init__.py:49]: This suggestion makes a lot of sense to me. I've seen localized imports before, but didn't really have a sense of their purpose. Keeping the imports of optional dependencies restricted to the functions that need them seems like an excellent approach.

File: pycytominer/cyto_utils/__init__.py
Comment (kenibrewer) [reply pycytominer/cyto_utils/__init__.py]: I had a hard time figuring out what kind of comment would be appropriate here based on your suggestion. We're not doing any other library imports in __init__.py so I'm not sure whether this a place someone would ever come looking for that type of information.

Comment (d33bs) [reply pycytominer/cyto_utils/__init__.py]: Thanks @kenibrewer , I think the changes you've added look great for now. Besides the changes you added, generally I wonder if a short GitHub Actions test for the installation and import of Pycytominer might avoid potential issues with the conda package testing (in advance of recipe updates). This might be overdoing it but could help avoid issues in the future.

**Review by d33bs — APPROVED** [review]
Thanks @kenibrewer - LGTM!

**Pull request merged**
